program/src/processor.rs

```diff
@@ -375,6 +376,101 @@ pub fn process_unwrap(accounts: &[AccountInfo], amount: u64) -> ProgramResult {
     Ok(())
 }
 
+/// Processes [`CloseStuckEscrow`](enum.TokenWrapInstruction.html) instruction.
+pub fn process_close_stuck_escrow(accounts: &[AccountInfo]) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+
+    let escrow_account = next_account_info(account_info_iter)?;
+    let destination_account = next_account_info(account_info_iter)?;
+    let unwrapped_mint = next_account_info(account_info_iter)?;
+    let wrapped_mint = next_account_info(account_info_iter)?;
+    let wrapped_mint_authority = next_account_info(account_info_iter)?;
+    let _token_2022_program = next_account_info(account_info_iter)?;
+
+    // This instruction is only for spl-token-2022 accounts because only they
+    // can have extensions that lead to size changes.
+    if *escrow_account.owner != spl_token_2022::id()
+        || unwrapped_mint.owner != &spl_token_2022::id()
+    {
+        return Err(ProgramError::IncorrectProgramId);
+    }
+
+    let expected_wrapped_mint_pubkey =
+        get_wrapped_mint_address(unwrapped_mint.key, wrapped_mint.owner);
+    if *wrapped_mint.key != expected_wrapped_mint_pubkey {
+        Err(TokenWrapError::WrappedMintMismatch)?
+    }
+
+    let (expected_authority, bump) = get_wrapped_mint_authority_with_seed(wrapped_mint.key);
+    if *wrapped_mint_authority.key != expected_authority {
+        Err(TokenWrapError::MintAuthorityMismatch)?
+    }
+
+    let expected_escrow_address = get_associated_token_address_with_program_id(
+        wrapped_mint_authority.key,
+        unwrapped_mint.key,
+        unwrapped_mint.owner,
+    );
+
+    if *escrow_account.key != expected_escrow_address {
+        return Err(TokenWrapError::EscrowMismatch.into());
+    }
+
+    let escrow_data = escrow_account.try_borrow_data()?;
+    let escrow_state = PodStateWithExtensions::<PodAccount>::unpack(&escrow_data)?;
+
+    if escrow_state.base.owner != *wrapped_mint_authority.key {
+        return Err(TokenWrapError::EscrowOwnerMismatch.into());
+    }
+
+    // Closing only works when the token balance is zero
+    if u64::from(escrow_state.base.amount) != 0 {
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    let current_account_extensions = escrow_state.get_extension_types()?;
+    drop(escrow_data);
+
+    let mint_data = unwrapped_mint.try_borrow_data()?;
+    let mint_state = PodStateWithExtensions::<PodMint>::unpack(&mint_data)?;
+    let mint_extensions = mint_state.get_extension_types()?;
+    let required_account_extensions =
+        ExtensionType::get_required_init_account_extensions(&mint_extensions);
+
+    // If the token account already shares the same extensions as the mint,
+    // it does not need to be re-created
+    let in_good_state = current_account_extensions.len() == required_account_extensions.len()
+        && required_account_extensions
+            .iter()
+            .all(|item| current_account_extensions.contains(item));
```

Comment:
This check won't be correct unfortunately -- the ATA program adds the `ImmutableOwner` extension, which is not required by any mint, so we'll also need that extension in `required_account_extensions`.

Reply:
! 

Thanks for catching this. I was surprised this slipped by my tests and realized my test setup is faulty as it does not include that `ImmutableOwner` extension. Will fix 🙏